src/arch/metal.rs

```diff
@@ -769,8 +770,15 @@ stack_fill_loop:
 stack_fill_done:
 
     // Now we need to zero-out the BSS section
-    // TODO: this is currently done by all cores, but we should have some sort of synchronization
-    // to avoid race conditions.
+    // Only the boot hart set the BSS section to avoid race condition.
+
+    csrr t0, mhartid       // Our current hart ID
+    ld t2, {boot_hart_id}  // Boot hart ID
+    bne t0, t2, wait_bss_end // Only the boot hart initializes the bss
+    
+    // Shared boolean, set to 1 to say to other harts that the BSS is not initialized yet
+    ld t3, __boot_bss_set
```

Comment:
You load the address of the boolean _after_ the jump of the non-boot harts. So the non-boot harts are truing to read from `t3` but `t3` might contain some invalid address.

You should load it before to ensure all threads have the address of the variable in `t3`.